File: step3_migration_to_kotlin/src/main/kotlin/java/migration/simple/service/StatsService.java
package migration.simple.service;

import migration.simple.repository.UserRepository;
import migration.simple.types.Stats;
import migration.simple.types.User;
import org.springframework.stereotype.Service;

import java.util.Comparator;
import java.util.List;

@Service
public class StatsService {

    private UserRepository userRepository;

    public StatsService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public Stats getStats() {
        List<User> allUsers = userRepository.findAllUsers();

        User oldestUser = allUsers.stream()
                .max(Comparator.comparingInt(User::getAge))
                .get();

        User youngestUser = allUsers.stream()
                .min(Comparator.comparingInt(User::getAge))
                .get();

        return new Stats(
                allUsers.size(),
                oldestUser,
                youngestUser
        );
    }
}

File: step4_migration_to_webflux/src/main/kotlin/migration/simple/service/StatsService.kt
package migration.simple.service

import migration.simple.repository.UserRepository
import migration.simple.types.Stats

open class StatsService(private val userRepository: UserRepository) {

    open fun getStats(): Stats {
        val allUsers = userRepository.findAllUsers()

        if (allUsers.isEmpty())
            throw RuntimeException("not find any user")

        val oldestUser = allUsers.maxBy { it.age }

        val youngestUser = allUsers.minBy { it.age }

        return Stats(
                allUsers.size,
                oldestUser!!,
                youngestUser!!
        )
    }
}
